src/libs/sfdk/buildengine.cpp

```diff
@@ -0,0 +1,1133 @@
+/****************************************************************************
+**
+** Copyright (C) 2012-2019 Jolla Ltd.
+** Copyright (C) 2019 Open Mobile Platform LLC.
+** Contact: http://jolla.com/
+**
+** This file is part of Qt Creator.
+**
+** GNU Lesser General Public License Usage
+** Alternatively, this file may be used under the terms of the GNU Lesser
+** General Public License version 2.1 as published by the Free Software
+** Foundation and appearing in the file LICENSE.LGPL included in the
+** packaging of this file.  Please review the following information to
+** ensure the GNU Lesser General Public License version 2.1 requirements
+** will be met: http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.
+**
+** Other Usage
+**
+** Alternatively, this file may be used in accordance with the terms and
+** conditions contained in a signed written agreement between you and Digia.
+**
+****************************************************************************/
+
+#include "buildengine_p.h"
+
+#include "sfdkconstants.h"
+#include "sdk_p.h"
+#include "targetsxmlreader_p.h"
+#include "usersettings_p.h"
+#include "virtualboxmanager_p.h"
+#include "vboxvirtualmachine_p.h"
+
+#include <ssh/sshconnection.h>
+#include <utils/algorithm.h>
+#include <utils/filesystemwatcher.h>
+#include <utils/persistentsettings.h>
+#include <utils/pointeralgorithm.h>
+#include <utils/qtcassert.h>
+
+#include <QDir>
+#include <QPointer>
+#include <QProcess>
+
+using namespace QSsh;
+using namespace Utils;
+
+namespace Sfdk {
+
+namespace {
+
+const char* SIMPLE_WRAPPERS[] = {
+    Constants::WRAPPER_QMAKE,
+    Constants::WRAPPER_MAKE,
+    Constants::WRAPPER_GCC,
+    Constants::WRAPPER_DEPLOY,
+    Constants::WRAPPER_RPM,
+    Constants::WRAPPER_RPMVALIDATION,
+    Constants::WRAPPER_LUPDATE,
+};
+
+} // namespace anonymous
+
+/*!
+ * \class RpmValidationSuiteData
+ */
+
+/*!
+ * \class BuildTargetData
+ */
+
+bool BuildTargetData::isValid() const
+{
+    return !name.isEmpty()
+        && !sysRoot.isEmpty()
+        && !toolsPath.isEmpty()
+        && !gdb.isEmpty();
+}
+
+Utils::FileName BuildTargetData::toolsPathCommonPrefix()
+{
+    return SdkPrivate::settingsLocation(SdkPrivate::UserScope)
+        .appendPath(Constants::BUILD_TARGET_TOOLS);
+}
+
+/*!
+ * \class BuildEngine::PrivateConstructorTag
+ * \internal
+ */
+
+struct BuildEngine::PrivateConstructorTag {};
```

Comment:
What's the point of PrivateConstructorTag? What improvement is achieved by having to call BuildEngine(parent, PrivateConstructorTag{}) instead of just BuildEngine(parent)?

Reply:
https://stackoverflow.com/questions/32351296/factory-pattern-with-private-constructors-in-c

Reply:
Or better https://stackoverflow.com/questions/8147027/how-do-i-call-stdmake-shared-on-a-class-with-only-protected-or-private-const

Reply:
That's what I thought this code tried to accomplish. 
But in our case PrivateConstructorTag::PrivateConstructorTag() is public, so anyone can just call BuildEngine(parent, PrivateConstructorTag{}). 
In your first link the tag is private (with a friend Factory class), and in the second it's protected, thus making it impossible to create it in unwanted places. 
Am I still missing something here?

Reply:
yup, the declaration is public, but the definition is private. Declaring the manager class as a friend to the tag would improve this further, replacing linker error with compiler error, but I do not try to overcomplicate access control at package scope - if it has private or internal in its name, it should be clear enough it shouldn't be used from outside.

Reply:
Sorry I keep bugging you about this, it's just because I have this feeling that there is more here than meets (my) eye. What would happen if we removed PrivateConstructorTag completely? Would some std::whatever() function somewhere generate extra BuildEngines? Or is it really just to make it more difficult (for the human programmer) to create BuildEngines (and possibly bugs)?

Reply:
It is there 1) to make it _impossible_ for the outside programmer to create build engines other way than via Sdk::createBuildEngine() and 2) to make the internal programmer stop and think before they try to create a build engine, because it needs some special care from the BuildEngineManager side https://github.com/sailfishos/sailfish-qtcreator/blob/jb45560/src/libs/sfdk/buildengine.cpp#L961